### ISSUE TITLE
Build error with Clang 14 from ROCm 5.0

### ISSUE THREAD

**Issue by nordmoen** [AdaptiveCpp/AdaptiveCpp#709]
**Bug summary**
When building with Clang 14 from ROCm 5.0 I encounter an error that `error: no member named 'SROAPass' in namespace 'llvm'` for `PipelineBuilder.cpp:87:23`. I have tried with `-DHIPSYCL_NO_DEVICE_MANGLER=ON`, but it did not help compilation.

**To Reproduce**
* Install EasyBuild
* Download ROCm version 5.0 (easybuilders/easybuild-easyconfigs#14995)
* Use this ROCm version to build `hipSYCL`

**Describe your setup**
* Trying to build hipSYCL `0.9.2`
* ROCm `5.0.0`
* Clang `14.0.0` 

**Additional context**
Potential missing functionality in ROCm Clang distribution?

EasyBuild receipt to build `hipSYCL`:
```python
easyblock = 'CMakeMake'

name = 'hipSYCL'
version = '0.9.2'
local_rocm_ver = '5.0.0'
versionsuffix = '-ROCm-%s' % local_rocm_ver

homepage = 'https://github.com/illuhad/hipSYCL'
description = """hipSYCL is a modern SYCL implementation targeting CPUs and
GPUs, with a focus on leveraging existing toolchains such as CUDA or HIP"""

toolchain = {'name': 'GCC', 'version': '11.2.0'}

source_urls = ["https://github.com/illuhad/hipSYCL/archive/refs/tags/"]
sources = ["v%(version)s.tar.gz"]
checksums = ['086cd259345627cd1205a2fbb96fb1f3d1c4f3a13f9520753a4f4df313445c6d']

dependencies = [
    ('Boost', '1.77.0'),
    ('Python', '3.9.6'),
    ('ROCm', local_rocm_ver),
]

builddependencies = [
    ('CMake', '3.22.1'),
]

configopts = '-DCMAKE_C_COMPILER="$EBROOTCLANGMINAOMP/bin/clang" '
configopts += '-DCMAKE_CXX_COMPILER="$EBROOTCLANGMINAOMP/bin/clang++" '
configopts += '-DROCM_PATH=$ROCM_PATH'
configopts += '-DWITH_CPU_BACKEND=ON '
configopts += '-DWITH_CUDA_BACKEND=OFF '
configopts += '-DWITH_ROCM_BACKEND=ON '
# The following option is needed with ROCm since the ROCm Clang distribution
# does not support multipass support
# https://github.com/illuhad/hipSYCL/blob/develop/doc/install-rocm.md
configopts += '-DHIPSYCL_NO_DEVICE_MANGLER=ON '

sanity_check_paths = {
    'files': ['bin/syclcc-clang', 'include/sycl/sycl.hpp',
              'lib/hipSYCL/librt-backend-omp.%s' % SHLIB_EXT],
    'dirs': ['include/CL', 'include/hipSYCL', 'include/SYCL', 'lib'],
}
sanity_check_commands = ['syclcc --help']

moduleclass = 'compiler'
```

**Comment by fodinabor** [comment]
Hey,
as commented in the newly created issue on ROCm side (https://github.com/RadeonOpenCompute/ROCm/issues/1680), the issue here is that the LLVM version ROCm uses is not really equal to the LLVM 14 version the Clang plugin expects.

If from the ROCm team no further advice follows, we see two major options in solving this, providing you another workaround macro (likely more of the form of `-DROCM_LLVM_VERSION=050000` or similar) or trying to detect the ROCm Clang's version the plugin is built against.

I guess providing the ROCm version as an additional configuration macro would be acceptable in your case?

As a temporary workaround, you could disable building the accelerated CPU support by adding `-DWITH_ACCELERATED_CPU=false` to the `configopts` although that'd likely be a bummer, if that stayed that way.

**Comment by nordmoen** [comment]
Thanks for the quick response @fodinabor and for working on solving this issue!

Building hipSYCL indeed works with `-DWITH_ACCELERATED_CPU=OFF`, so that is a workaround for now.

**Comment by illuhad** [comment]
I'd like to add that you probably don't have to use `DHIPSYCL_NO_DEVICE_MANGLER=ON`. This is a workaround for a similar issue with clang "13" that ships with ROCm 4.5 not really having all the functionality of clang 13. Since ROCm 5 is based on the clang 14 branch and is hence post clang 13 release, this should no longer be an issue.

Be advised that with the `-DWITH_ACCELERATED_CPU=OFF` workaround performance of the CPU backend will likely be several orders of magnitude slower for `nd_range` kernels containing barriers.

Another solution would be to just not build against ROCm clang, but against upstream clang 14. This might also work with ROCm 5. As stated in the documentation, building against ROCm clang is only semi-supported, and the more clang/LLVM functionality hipSYCL relies on, the harder will this be to support. The core issue really is that AMD branches off their ROCm clang whenever they want, and then don't merge in upstream changes.

**Comment by Maxzor** [comment]
They upstream their changes. It is just hard to do it in a timely manner when both LLVM and ROCm releases are not synced, with the branching problem that you mentioned, it is the root of many issues.

We are at Debian using upstream Clang 14 rc1 for ROCm 5.0 and it sort of works.
https://github.com/ROCm-Developer-Tools/HIP/issues/2449
https://lists.debian.org/debian-ai/2022/02/msg00047.html
https://salsa.debian.org/rocm-team/community/team-project#packaging-progress
AMD, having opened its backend and forked LLVM, is both a blessing and a curse.

**Comment by illuhad** [comment]
@Maxzor Yeah, in my experience the only real issue with the combination of upstream clang and ROCm is if you want to run brand-new ROCm which might require clang patches that have not yet been upstreamed, or need a very new compiler because you have a brand-new GPU. Apart from that we've never had issues with upstream clang in hipSYCL.
We use upstream clang in our ROCm CI, not ROCm clang, and never had issues with it. So, for hipSYCL upstream clang on AMD GPUs is definitely better tested and more stable.

I think when using something like OpenMP offloading the situation might be different, but hipSYCL does not need these features anyway.
